v03_pipeline/lib/misc/allele_registry.py

```diff
@@ -0,0 +1,145 @@
+        else:
+            chunk_ht = ht.tail(end_idx - num_rows)
+        yield register_alleles(chunk_ht, reference_genome, base_url)
+
+
+def register_alleles(
+    ht: hl.Table,
+    reference_genome: ReferenceGenome,
+    base_url: str,
+) -> dict[str, str]:
+    with tempfile.NamedTemporaryFile(
+        suffix='.vcf',
+    ) as raw_vcf, tempfile.NamedTemporaryFile(suffix='.vcf') as formatted_vcf:
+        hl.export_vcf(ht, raw_vcf.name)
+
+        # Reformat the VCF created by hail's 'export_vcf' function: replace the header and remove any 'chr' prefix
+        with open(raw_vcf.name) as vcf_in, open(formatted_vcf.name, 'w') as vcf_out:
+            vcf_out.writelines(reference_genome.allele_registry_vcf_header)
+            for line in vcf_in:
+                if not line.startswith('#'):
+                    vcf_out.write(line.replace('chr', ''))
+
+        logger.info('Calling the Clingen Allele Registry.')
+        with open(formatted_vcf.name) as vcf_in:
+            data = vcf_in.read()
+            res = requests.put(
+                url=build_url(base_url),
+                data=data,
+                timeout=HTTP_REQUEST_TIMEOUT,
+            )
+            return handle_api_response(res, base_url)
+
+
+def build_url(base_url: str) -> str:
+    login = Env.ALLELE_REGISTRY_LOGIN
+    password = Env.ALLELE_REGISTRY_PASSWORD
+
+    if login is None or password is None:
+        msg = 'Please set the ALLELE_REGISTRY_LOGIN and ALLELE_REGISTRY_PASSWORD environment variables.'
+        raise ValueError(msg)
+
+    # adapted from https://reg.clinicalgenome.org/doc/scripts/request_with_payload.py
+    identity = hashlib.sha1((login + password).encode('utf-8')).hexdigest()  # noqa: S324
+    gb_time = str(int(time.time()))
+    token = hashlib.sha1((base_url + identity + gb_time).encode('utf-8')).hexdigest()  # noqa: S324
+    return base_url + '&gbLogin=' + login + '&gbTime=' + gb_time + '&gbToken=' + token
+
+
+def handle_api_response(res: requests.Response, base_url: str) -> dict[str, str]:
+    response = res.json()
+    if not res.ok or 'errorType' in response:
+        error = AlleleRegistryError.from_api_response(response, base_url)
+        logger.error(error.loggable_message)
+        raise HTTPError(error.message)
+
+    id_map = {}
+    errors = []
+    for allele_response in response:
+        if 'errorType' in allele_response:
+            errors.append(
+                AlleleRegistryError.from_api_response(allele_response, base_url),
+            )
+        else:
+            # Example allele_response:
+            # {'@id': 'http://reg.genome.network/allele/CA520798109',
+            #  'externalRecords': {'gnomAD_4': [{'id': '1-10109-AACCCT-A'}]}}
+            caid = allele_response['@id'].split('/')[-1]
+            if 'externalRecords' in allele_response:
+                gnomad_id = allele_response['externalRecords']['gnomAD_4'][0]['id']
+                id_map[gnomad_id] = caid
```

Comment:
I'm currently using gnomad_id from the api response to map to CAID when [annotating](https://github.com/broadinstitute/seqr-loading-pipelines/pull/759/files#diff-2677a48e0bab4da76f54ec6184fa2179f8a688a291a46faeaf012eef58e74274R248) new variants with CAID. 

However not all variants have a gnomad ID, some have other IDs. I'm thinking of alternatives/better ways to do this mapping.

Attached is an example allele registry json response with no fields filtered for this variant: 1:10109	["AACCCT","A"]
[allele_reg_json_response_example.json](https://github.com/broadinstitute/seqr-loading-pipelines/files/15030753/allele_reg_json_response_example.json)

Reply:
So in seqr we have this line of code that takes a a gnomad style ID thats been split on a "-" (literally called `variant_id.split('-')` upstream somewhere) and turns it into a locus and allele, which should be the key of the table you are working with. You could then turn the allele response into its own table keyed by locus-allele so you can join them
https://github.com/broadinstitute/seqr/blob/master/hail_search/queries/mito.py#L202

Reply:
I like this approach, (left) joining two hail tables to annotate CAID definitely feels cleaner than annotating a new field from a dictionary.  Performance-wise I'm guessing the join is faster this way too?